File: dashboard/src/contract.ts
// Mirror of src/ga/contract.py (PRD §3). Python is canonical; these literals must
// match it (test_contract.py::test_enums_match_frozen_spec pins that side).

export type Frame = "direct" | "roleplay" | "evaluation" | "fictional" | "document_analysis";
export type Persona = "none" | "auditor" | "developer" | "teacher" | "assistant" | "character";
export type TaskStyle = "answer" | "summarize" | "classify" | "translate" | "complete" | "rewrite";
export type InstructionPressure = "low" | "medium" | "high";
export type ConversationMode = "single_turn" | "multi_turn";
export type ContextSource = "direct_user_prompt" | "quoted_text" | "external_document";

export type Format = "plain" | "markdown" | "json" | "xml" | "table" | "dialogue";
export type DelimiterStyle = "none" | "quotes" | "code_block" | "markdown_headers";
export type NoiseType = "none" | "random_chars" | "spacing" | "casing" | "typo" | "encoding_like";
export type NoisePosition = "prefix" | "suffix" | "interleaved" | "around_target";

export type Outcome = "violation" | "partial" | "refusal" | "malformed";
export type Search = "genetic" | "random";

export interface SemanticChannel {
  frame: Frame;
  persona: Persona;
  task_style: TaskStyle;
  instruction_pressure: InstructionPressure;
  demo_count: number;
  conversation_mode: ConversationMode;
  context_source: ContextSource;
}

export interface PerturbationChannel {
  format: Format;
  delimiter_style: DelimiterStyle;
  noise_enabled: boolean;
  noise_type: NoiseType;
  noise_position: NoisePosition;
  noise_ratio: number;
}

export interface Genome {
  semantic_channel: SemanticChannel;
  perturbation_channel: PerturbationChannel;
}

export interface ResultRecord {
  run_id: string;
  search: Search;
  generation: number;
  genome_id: string;
  parent_ids: string[];
  genome: Genome;
  rendered_prompt: string;
  policy_id: string;
  target_query_id: string;
  response: string;
  outcome: Outcome;
  fitness: number;
  seed: number;
}

const OUTCOMES: ReadonlySet<string> = new Set(["violation", "partial", "refusal", "malformed"]);
const SEARCHES: ReadonlySet<string> = new Set(["genetic", "random"]);

// Boundary guard: an external file is untrusted, so validate shape before it flows
// inward. Cheap structural check of the load-bearing fields.
export function isResultRecord(x: unknown): x is ResultRecord {
  if (typeof x !== "object" || x === null) return false;
  const r = x as Record<string, unknown>;
  return (
    SEARCHES.has(r.search as string) &&
    OUTCOMES.has(r.outcome as string) &&
    typeof r.generation === "number" &&
    typeof r.genome_id === "string" &&
    Array.isArray(r.parent_ids) &&
    typeof r.fitness === "number" &&
    typeof r.genome === "object" &&
    r.genome !== null &&
    typeof (r.genome as Record<string, unknown>).semantic_channel === "object" &&
    typeof (r.genome as Record<string, unknown>).perturbation_channel === "object"
  );
}

File: src/ga/contract.py
import hashlib
import json
from typing import Literal, get_args

from pydantic import BaseModel, ConfigDict, Field

# Boundary contract (PRD §3) — the single source of truth at the Python↔TS seam.
# The TS mirror lives in dashboard/src/contract.ts and must match these literals;
# test_contract.py::test_enums_match_frozen_spec pins this side so it can't drift.

Frame = Literal["direct", "roleplay", "evaluation", "fictional", "document_analysis"]
Persona = Literal["none", "auditor", "developer", "teacher", "assistant", "character"]
TaskStyle = Literal["answer", "summarize", "classify", "translate", "complete", "rewrite"]
InstructionPressure = Literal["low", "medium", "high"]
ConversationMode = Literal["single_turn", "multi_turn"]
ContextSource = Literal["direct_user_prompt", "quoted_text", "external_document"]

Format = Literal["plain", "markdown", "json", "xml", "table", "dialogue"]
DelimiterStyle = Literal["none", "quotes", "code_block", "markdown_headers"]
NoiseType = Literal["none", "random_chars", "spacing", "casing", "typo", "encoding_like"]
NoisePosition = Literal["prefix", "suffix", "interleaved", "around_target"]

Outcome = Literal["violation", "partial", "refusal", "malformed"]
Search = Literal["genetic", "random"]


class SemanticChannel(BaseModel):
    model_config = ConfigDict(extra="forbid")
    frame: Frame
    persona: Persona
    task_style: TaskStyle
    instruction_pressure: InstructionPressure
    demo_count: int = Field(ge=0, le=5)
    conversation_mode: ConversationMode
    context_source: ContextSource


class PerturbationChannel(BaseModel):
    model_config = ConfigDict(extra="forbid")
    format: Format
    delimiter_style: DelimiterStyle
    noise_enabled: bool
    noise_type: NoiseType
    noise_position: NoisePosition
    noise_ratio: float = Field(ge=0.0, le=1.0)


class Genome(BaseModel):
    model_config = ConfigDict(extra="forbid")
    semantic_channel: SemanticChannel
    perturbation_channel: PerturbationChannel

    @property
    def genome_id(self) -> str:
        canonical = json.dumps(self.model_dump(), sort_keys=True, separators=(",", ":"))
        return hashlib.sha1(canonical.encode()).hexdigest()[:12]


class ResultRecord(BaseModel):
    model_config = ConfigDict(extra="forbid")
    run_id: str
    search: Search
    generation: int = Field(ge=0)
    genome_id: str
    parent_ids: list[str]
    genome: Genome
    rendered_prompt: str
    policy_id: str
    target_query_id: str
    response: str
    outcome: Outcome
    fitness: float = Field(ge=0.0, le=1.0)
    seed: int


# Categorical gene domains, derived from the Literal annotations above so sampling
# and mutation can never disagree with validation. demo_count / noise_ratio /
# noise_enabled are numeric/bool and handled directly by the operators.
def categorical_values() -> tuple[dict[str, list], dict[str, list]]:
    def vals(model: type[BaseModel], field: str) -> list:
        return list(get_args(model.model_fields[field].annotation))

    semantic = {
        f: vals(SemanticChannel, f)
        for f in ("frame", "persona", "task_style", "instruction_pressure", "conversation_mode", "context_source")
    }
    perturbation = {
        f: vals(PerturbationChannel, f)
        for f in ("format", "delimiter_style", "noise_type", "noise_position")
    }
    return semantic, perturbation
